File: 39.combination-sum.java
/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    private List<Integer> combie;
    private Set<List<Integer>> results;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        combie = new ArrayList<>();
        results = new HashSet<>();
        backtracking(candidates, target);
        return new ArrayList<>(results);
    }
    private void backtracking(int[] candidates, int target) {
        int sum = combie.stream().mapToInt(x -> x).sum();
        if (sum > target) return;
        if (sum == target) {
            List<Integer> tmp = new ArrayList<>(combie);
            Collections.sort(tmp);
            results.add(new ArrayList<>(tmp));
            return;
        }
        for (int n : candidates) {
            combie.add(n);
            backtracking(candidates, target);
            combie.removeLast();
        }
    }
}
// @lc code=end

File: 39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */

// @lc code=start

#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
private:
    vector<int> candidates_;
    int target_;
    vector<vector<int>> results_;
    vector<int> bag_;
    int sum_ {};
    void backtracking(vector<int>::iterator begin) {
        if (sum_ == target_) {
            results_.push_back(bag_);
            return;
        }
        if (sum_ > target_) return;
        for (auto iter = begin; iter != candidates_.end(); ++iter) {
            bag_.push_back(*iter);
            sum_ += *iter;
            backtracking(iter);
            bag_.pop_back();
            sum_ -= *iter;
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        candidates_ = candidates;
        target_ = target;
        backtracking(candidates_.begin());
        return results_;
    }

};
// @lc code=end
